File: Python/687.longest-univalue-path.py
#
# @lc app=leetcode id=687 lang=python
#
# [687] Longest Univalue Path
#
# https://leetcode.com/problems/longest-univalue-path/description/
#
# algorithms
# Easy (34.64%)
# Likes:    1207
# Dislikes: 314
# Total Accepted:    71.7K
# Total Submissions: 206.6K
# Testcase Example:  '[5,4,5,1,1,5]'
#
# Given a binary tree, find the length of the longest path where each node in
# the path has the same value. This path may or may not pass through the root.
# 
# The length of path between two nodes is represented by the number of edges
# between them.
# 
# 
# 
# Example 1:
# 
# Input:
# 
# 
# ⁠             5
# ⁠            / \
# ⁠           4   5
# ⁠          / \   \
# ⁠         1   1   5
# 
# 
# Output: 2
# 
# 
# 
# Example 2:
# 
# Input:
# 
# 
# ⁠             1
# ⁠            / \
# ⁠           4   5
# ⁠          / \   \
# ⁠         4   4   5
# 
# 
# Output: 2
# 
# 
# 
# Note: The given binary tree has not more than 10000 nodes. The height of the
# tree is not more than 1000.
# 
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def longestUnivaluePath(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """

    #     self.ans = 0
    #     self.dfs(root)
    #     return self.ans

    # def dfs(self, root):
    #     if not root:
    #         return 0

    #     # 先计算，不然会覆盖结果
    #     left = self.dfs(root.left)
    #     right = self.dfs(root.right)

    #     if root.left and root.left.val == root.val:
    #         left_path = left+1  
    #     else:
    #         left_path = 0
    #     if root.right and root.right.val == root.val:
    #         right_path = right+1  
    #     else:
    #         right_path = 0
    #     self.ans = max(self.ans, left_path + right_path)
    #     return max(left_path, right_path)


        return self.dfs(root)[0]
    def dfs(self, root):
        if not root:
            return (0, 0)
        l1, l2 = self.dfs(root.left)
        r1, r2 = self.dfs(root.right)

        if root.left and root.left.val == root.val:
            l2 += 1
        else:
            l2 = 0
        if root.right and root.right.val == root.val:
            r2 += 1
        else:
            r2 = 0 
        return max(l1, r1, l2 + r2), max(l2, r2)


# @lc code=end

sol = Solution()
root = TreeNode(5)
root.left = TreeNode(4) 
root.left.left = TreeNode(1)
root.left.right = TreeNode(1)
root.right = TreeNode(5)
root.right.right = TreeNode(5)
print(sol.longestUnivaluePath(root))

File: C++/687.longest-univalue-path.cpp
/*
 * @lc app=leetcode id=687 lang=cpp
 *
 * [687] Longest Univalue Path
 *
 * https://leetcode.com/problems/longest-univalue-path/description/
 *
 * algorithms
 * Easy (35.08%)
 * Likes:    1330
 * Dislikes: 356
 * Total Accepted:    77.3K
 * Total Submissions: 220.4K
 * Testcase Example:  '[5,4,5,1,1,5]'
 *
 * Given a binary tree, find the length of the longest path where each node in
 * the path has the same value. This path may or may not pass through the
 * root.
 * 
 * The length of path between two nodes is represented by the number of edges
 * between them.
 * 
 * 
 * 
 * Example 1:
 * 
 * Input:
 * 
 * 
 * ⁠             5
 * ⁠            / \
 * ⁠           4   5
 * ⁠          / \   \
 * ⁠         1   1   5
 * 
 * 
 * Output: 2
 * 
 * 
 * 
 * Example 2:
 * 
 * Input:
 * 
 * 
 * ⁠             1
 * ⁠            / \
 * ⁠           4   5
 * ⁠          / \   \
 * ⁠         4   4   5
 * 
 * 
 * Output: 2
 * 
 * 
 * 
 * Note: The given binary tree has not more than 10000 nodes. The height of the
 * tree is not more than 1000.
 * 
 */
#include<vector>
#include<unordered_map>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
// @lc code=start
class Solution {
public:

    // Case 1: If the root value is equal to BOTH the left and right subtree root values, then set m to the maximum of itself compared to the "up-side-down-V shaped" path created by including the left and right subtrees max paths with the root ( l=l+1 and r=r+1 ). Return the max path value of the left and right subtrees return max(l,r) which is needed in case the next parent up the recursive stack is also the same value.

    // Case 2: If the root value is NOT equal to EITHER the left INCLUSIVE-OR right subtree values, then set BOTH l=0 and r=0 and return 0 ( i.e. return max(0,0) ).

    // Case 3: If the root value is equal to the left subtree root value XOR the right subtree root value, then either l=0 and r=r+1 XOR l=l+1 and r=0, return the maximum of the left and right subtree path values ( i.e. return max(l,0) XOR return max(0,r) ).

    int longestUnivaluePath(TreeNode* root) {
        int ans = 0;
        DFS(root, ans);
        return ans;
    }

    int DFS(TreeNode* node, int& ans) {
        if (node==NULL) return 0;
        int tmp_left = DFS(node->left, ans);
        int tmp_right = DFS(node->right, ans);

        int left_path = 0, right_path = 0;
        if (node->left && node->left->val==node->val) left_path = tmp_left+1;
        if (node->right && node->right->val==node->val) right_path = tmp_right+1;
        ans = max(ans, left_path + right_path);
        return max(left_path, right_path);
    }
};
// @lc code=end

